### ISSUE TITLE
Compiler Performance: Lazily construct SourceFile objects from input files

### ISSUE THREAD

**Issue by DavidANeil** [microsoft/TypeScript#37701]
## Motivation

When constructing a ts.Program the following line of code runs
```typescript
forEach(rootNames, name => processRootFile(name, /*isDefaultLib*/ false, /*ignoreNoDefaultLib*/ false));
```

Performance tracing is reporting this single loop as taking anywhere between about 30-80% of the time it takes to type check and transpile our TS.

Even with providing a host that has cached the source text of most of those `rootNames` files in memory, and also cached the `ts.SourceFile` object, this loop takes about 1ms per file in `rootNames`, which isn't very significant alone, but it adds up quickly when you consider that our full build involves ~16,000 source files in ~1500 programs with a grand total of ~6,000,000 `ts.SourceFile` cache reads.

In some cases we are able to just reuse an `oldProgram` which results in a fairly drastic speed up, but that solution can't be used on our CI.

## Proposal

Through examination of the files it is apparent that most of the files that are inputs are not actually used for any compiler actions (i.e. for type checking).  As such it would be beneficial to avoid loading all of the files during initialization and instead load them lazily as needed.

## Global Namespaces

The obvious issue with not opening every single file is that some declarations have side-effects, such as module augmentation and global namespaces.  I'd propose that such files be passed to the compiler through a different mechanism than the `files` attribute, (such as `globalFiles`) to denote that they will be treated differently by the compiler.

## Extra Benefits

Under Bazel, if an input file is never read it can be safely considered as "unused".  Lazily reading the files as needed would allow Bazel to not invalidate its build cache when a file that doesn't affect the output is changed.

**Comment by RyanCavanaugh** [comment]
Observation: This loop is also about expanding the set of files in the program - files can contain `import`s, `<reference path`, or `<reference types` directives that result in more files being added to the program. The only way this mode would work is if the program also built correctly under `--noResolve`, which is somewhat rare (though less rare due to `composite` enforcing *some* of these restrictions).

**Comment by RyanCavanaugh** [comment]
cc @sheetalkamat for feedback as well

**Comment by sheetalkamat** [comment]
Agreed with @RyanCavanaugh that source files are need to know the world to give correct information. The sourceFiles parsing doesnt take long in my opinion.. Its resolving the modules/type references/ triple slash references and then doing that recursively on those files is what causes the slowness of that loop. The theory is that with --build we should be creating fewer programs (that is only when build is really needed) and in those circumstances, we do need whole world so as @RyanCavanaugh mentioned is it `--noResolve` that you are looking for that tells to not resolve all the imports?

**Comment by DavidANeil** [comment]
We already have `noResolve` enabled.

**Comment by evmar** [comment]
I have some context on this one.  We have observed the same phenomenon within Google.

Because bazel manages the inputs, rules_typescript uses noResolve, which means it effectively passes **/*.ts as 'files' to every compilation, regardless of whether they're needed.  It only type checks the direct sources to the bazel rule, but the transitive inputs still cost, as described in this bg.

I am close to landing a change to set noResolve back to the default, and instead pass only the current compilation's sources as files.  This means the compiler then loads the srcs and anything they transitively import, excluding any unneeded files.  In my experiments this reduced a sample build step from e.g. 7k files to 1k files, and reduced build time by up to 50% (depending on how many unneeded files there are).

As you observe, the hard part is what to do about files that 'declare global' -- you need these files to compile, but you don't want to parse these files while compiling (that's the whole point).  I've had to carefully work around those in the bazel guts.

My change will take another few weeks to land, and also it's totally undocumented, so I dunno what it means for rules_typescript.  But my comment on this bug is that the underlying issue here is an artifact of how rules_typescript constructs TypeScript builds, and not an issue with the TS compiler itself.

**Comment by pauldraper** [comment]
Don't you just need to remove noResolve and pass in the list of "root" files?

That would then create SourceFiles only as needed? (And then track them for your unused files thing as well?)

**Comment by DavidANeil** [comment]
Yes, that is exactly what we have done since Evan left that comment.  The real trick is that "root files" needs to include all ambient declarations, because if `@types/node/fs.d.ts` isn't already loaded then, to my knowledge, TSC has no way of knowing that it should look in said file when an `import ... from 'fs';` is encountered.

The choice is then between passing in those ambient declarations through the `files` attribute or through `typeRoots`.  Currently we are using `files`.

I am going to close this issue because tsc was doing exactly what it was told to do.

Mostly jokingly: It might be worth investigating if there is a way to tell someone they don't know what they're doing if they construct a program with many thousands of files and then only use compiler actions on a very small subset of them.
